### PR TITLE
Feature/upgrade to 9.3

### PR DESCRIPTION
**Pull request by praveenmanchana**
Upgrade the module to support 9.3 SXA
Added main.js which can be run as a local CriticalCSS generator. 
Refactored code to introduce DI and remove dependency on Event from command and tasks.

### CODE REVIEW

File: src/Foundation/Speedy/serialization/SpeedySettings/Speedy/Speedy Global Settings.yml [hunk 1]
```diff
@@ -7,26 +7,22 @@ DB: master
...
   Hint: RemoteFontMap
   Value: |
-    { "fontmap": [
+    { "fontMap": [
         {
-          "find": "../../local/Temp/fonts/national2web/",
-          "replace": "/-/media/Themes/RAA/Web/Web/fonts/national2web/"
+          "find": "/Users/sitecore/AppData/Local/fonts/tahoma/",
+          "replace": "/-/media/themes/anziif/anziif-theme/fonts/tahoma/"
         },
```
Comment (TomTyack) [inline src/Foundation/Speedy/serialization/SpeedySettings/Speedy/Speedy Global Settings.yml]: The font map is just an example it changes for each website dependent on what they use.

File: src/Foundation/Speedy/code/Services/CriticalGenerationLocal.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
+﻿using System;
+using Newtonsoft.Json;
+using RestSharp;
+using Sitecore.Foundation.Speedy.Model;
+using Sitecore.Foundation.Speedy.Settings;
+
+namespace Sitecore.Foundation.Speedy.Services
+{
+    public class CriticalGenerationLocal : ICriticalGenerationGateway
+    {
```
Comment (TomTyack) [inline src/Foundation/Speedy/code/Services/CriticalGenerationLocal.cs]: Might be some confusion here. This is using RestClient to call the remote API hosted on Azure. For production purposes.  It uses a remote API as it requires a none headless browser support which we can't get when node runs in Azure.

File: src/Foundation/Speedy/code/Settings/SpeedyGenerationSettings.cs [hunk 3]
```diff
@@ -86,23 +89,13 @@ public static bool IsCriticalJavascriptEnabledAndPossible(this Item item)
...
-        {
-            return GetContextDatabase().GetItem(SpeedyConstants.GlobalSettings.SpeedyGlobalSettingsId);
-        }
-
-        private static Item GetGlobalSettingsItemFromMaster()
         {
             return GetMasterDatabase().GetItem(SpeedyConstants.GlobalSettings.SpeedyGlobalSettingsId);
         }
 
-        private static Database GetContextDatabase()
```
Comment (TomTyack) [inline src/Foundation/Speedy/code/Settings/SpeedyGenerationSettings.cs]: The Get Context calls are required by the front end on both CD and CM to generate the new layout. Or is there a way around this ?

File: src/Foundation/Speedy/code/App_Config/Include/Speedy/Foundation/Foundation.Speedy.Settings.config [hunk 4]
```diff
@@ -1,10 +1,12 @@
...
-               xmlns:unicorn="http://www.sitecore.net/xmlconfig/unicorn/">
+               xmlns:role="http://www.sitecore.net/xmlconfig/role/">
   <sitecore>
     <settings>
       <setting name="Speedy.IsPublicFacingEnvironment" value="false" role:require="Standalone" />
-      <setting name="Speedy.IsPublicFacingEnvironment" value="true" role:require="ContentManagement OR ContentDelivery" />
+      <setting name="Speedy.IsPublicFacingEnvironment" value="true" role:require="!Standalone" />
+      <!--Update this to true if you are running local node critical css generator-->
+      <setting name="Speedy.UseLocalCriticalCssGenerator" value="false" role:require="Standalone" />
+      <setting name="Speedy.UseLocalCriticalCssGenerator" value="false" role:require="!Standalone" />
```
Comment (TomTyack) [inline src/Foundation/Speedy/code/App_Config/Include/Speedy/Foundation/Foundation.Speedy.Settings.config]: Double up on the same setting that negates itself?  Or am I reading this wrong?

File: src/Foundation/Speedy/serialization/SpeedySettings/Speedy/Speedy Global Settings.yml (hunk 1 above)
Comment (TomTyack) [reply src/Foundation/Speedy/serialization/SpeedySettings/Speedy/Speedy Global Settings.yml]: Ah unless this replacement matches the new habitat example.  That would make sense :)

Comment (TomTyack) [reply src/Foundation/Speedy/serialization/SpeedySettings/Speedy/Speedy Global Settings.yml]: Ah ok its for anziif ... we may want to roll that back for github.

File: src/Foundation/Speedy/serialization/SpeedySettings/Speedy/Speedy Global Settings.yml [hunk 5]
```diff
@@ -7,26 +7,22 @@ DB: master
...
         {
-          "find": "../../local/Temp/fonts/fontawesome/",
-          "replace": "/-/media/Themes/RAA/Web/Web/fonts/fontawesome/"
-        },
-        {
-          "find": "../../local/Temp/fonts/fontawesome/",
-          "replace": "/-/media/Themes/RAA/Web/Web/fonts/fontawesome/"
-        },
+          "find": "/Users/sitecore/AppData/Local/fonts/fontawesome/",
+          "replace": "/-/media/themes/anziif/anziif-theme/fonts/fontawesome/"
```
Comment (TomTyack) [inline src/Foundation/Speedy/serialization/SpeedySettings/Speedy/Speedy Global Settings.yml]: rollback Anzif

File: src/Foundation/Speedy/serialization/SpeedySettings/Speedy/Speedy Global Settings.yml [hunk 6]
```diff
@@ -82,21 +86,15 @@ Languages:
...
-                }
-            ]
+        "removeDuplicates": [
+            {
+              "find": "@font-face{font-display: swap;font-family:Conv_tahoma;src:url(/-/media/themes/anziif/anziif-theme/fonts/tahoma/tahoma_eot.eot);src:local(\"☺\"),url(/-/media/themes/anziif/anziif-theme/fonts/tahoma/tahoma_woff.woff) format(\"woff\"),url(/-/media/themes/anziif/anziif-theme/fonts/tahoma/tahoma_ttf.ttf) format(\"truetype\"),url(/-/media/themes/anziif/anziif-theme/fonts/tahoma/tahoma_svg.svg) format(\"svg\");font-weight:400;font-style:normal}"
+            },
+            {
+              "find": "@font-face{font-display: swap;src:url(/-/media/themes/anziif/anziif-theme/fonts/tahoma/tahoma_eot.eot);src:local(\"☺\"),url(/-/media/themes/anziif/anziif-theme/fonts/tahoma/tahoma_woff.woff) format(\"woff\"),url(/-/media/themes/anziif/anziif-theme/fonts/tahoma/tahoma_ttf.ttf) format(\"truetype\"),url(/-/media/themes/anziif/anziif-theme/fonts/tahoma/tahoma_svg.svg) format(\"svg\");font-weight:400;font-style:normal}"
```
Comment (TomTyack) [inline src/Foundation/Speedy/serialization/SpeedySettings/Speedy/Speedy Global Settings.yml]: Anzif

**Comment by TomTyack** [comment]
Confirmed its working on Habitat after the upgrade. main.js all good as well.  Just some minor questions.

File: src/Foundation/Speedy/code/Settings/SpeedyGenerationSettings.cs (hunk 3 above)
Comment (praveenmanchana) [reply src/Foundation/Speedy/code/Settings/SpeedyGenerationSettings.cs]: Currently all the Critical Generation is happening on CM. So we can save the item and then publish it? This is used to read the settings for API endpoint and usernames. And when running the Command(Ribbon) it seems to set the Context Database to Core which does not have these settings.

Comment (TomTyack) [reply src/Foundation/Speedy/code/Settings/SpeedyGenerationSettings.cs]: A few of the razor files used in the layout lookup a global settings item. Just when its coming from the page being generated on CD it requires the context database. 

Flow: 
Layout -> SpeedyJavascriptLoaderSettings.cshtml -> GetDeferJSLoadForMilliseconds() -> GetGlobalSettingsItem() -> GetContextDatabase()

**Pull request merged**